### PR TITLE
feat: Generic animation system

### PR DESCRIPTION
**Pull request by Rhuagh**
Taking a stab at making animation generic. 

- [X] Base types
- [X] Sampling system
- [x] Animation control system
- [X] LocalTransform added back in
- [x] GLTF loading
- [x] Examples

---
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/558)

### CODE REVIEW

**Comment by Rhuagh** [comment]
Ready for review.

**Comment by torkleyy** [comment]
This is great :)

---

Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[amethyst_animation/src/bundle.rs, line 16 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7MYHYfPRcs7h78f6:-L4Q7MYHYfPRcs7h78f7:bfukpl1) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L16)):*
> ```Rust
>               SamplerProcessor};
> 
> const DEP: [&str; 0] = [];
> ```

Why are you using a constant here? Isn't a reference to a ZST 'static automatically?

---

*[amethyst_animation/src/bundle.rs, line 59 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7v6RNQMsdO3KI7Fm:-L4Q7v6RNQMsdO3KI7Fn:b-oxc7ng) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L59)):*
> ```Rust
> /// Bundle for only the sampler interpolation.
> ///
> /// Will add `SamplerInterpolationSystem` with name `sampler_interpolation_system`.
> ```

`sampler_processor` isn't listed here

---

*[amethyst_animation/src/bundle.rs, line 93 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7dJWTmordw1cBMFg:-L4Q7dJWTmordw1cBMFh:bhhk2ki) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L93)):*
> ```Rust
>         builder: DispatcherBuilder<'a, 'b>,
>     ) -> Result<DispatcherBuilder<'a, 'b>> {
>         if !world
> ```

I think we have added an Entry API for this.

---

*[amethyst_animation/src/resources.rs, line 15 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q978GD0Jgpm7c6iaC:-L4Q978GD0Jgpm7c6iaD:brmwdav) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/resources.rs#L15)):*
> ```Rust
> pub trait AnimationSampling {
>     /// The channel type
>     type Channel;
> ```

Maybe add the `Send + Sync + 'static` bounds here?

---

*[amethyst_animation/src/resources.rs, line 23 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q8nlhENweM9hiU8J_:-L4Q8nlhENweM9hiU8Ja:bmxt0cl) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/resources.rs#L23)):*
> ```Rust
> 
>     /// Get the current sample for a channel
>     fn get_current_sample(&self, channel: &Self::Channel) -> SamplerPrimitive<Self::Scalar>;
> ```

Should be `current_sample`

---

*[amethyst_animation/src/resources.rs, line 264 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q9KKvHU8eKi04wFAJ:-L4Q9KKvHU8eKi04wFAK:bnsgacm) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/resources.rs#L264)):*
> ```Rust
>         command: AnimationCommand,
>     ) -> Self {
>         Self {
> ```

Please use `AnimationControl` for the construction

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558#-:-L4QA-j99fJZD1sihX85:bbmt9q3)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[amethyst_animation/src/bundle.rs, line 93 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7dJWTmordw1cBMFg:-L4QW-HX-C4sd55UCmhO:bk46tjx) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L93)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

I think we have added an Entry API for this.
</blockquote></details>

Released?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[amethyst_animation/src/bundle.rs, line 16 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7MYHYfPRcs7h78f6:-L4QXZaNnbTm5PIp7K59:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L16)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Why are you using a constant here? Isn't a reference to a ZST 'static automatically?
</blockquote></details>

Done.

---

*[amethyst_animation/src/bundle.rs, line 59 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7v6RNQMsdO3KI7Fm:-L4QX_1_kQ0RdWd1fuYL:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L59)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

`sampler_processor` isn't listed here
</blockquote></details>

Done.

---

*[amethyst_animation/src/resources.rs, line 15 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q978GD0Jgpm7c6iaC:-L4QX_uKk2Ld8GZSHRmb:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/resources.rs#L15)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Maybe add the `Send + Sync + 'static` bounds here?
</blockquote></details>

Done.

---

*[amethyst_animation/src/resources.rs, line 23 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q8nlhENweM9hiU8J_:-L4QXaHSEH4Pc-vjSi3P:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/resources.rs#L23)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Should be `current_sample`
</blockquote></details>

Done.

---

*[amethyst_animation/src/resources.rs, line 264 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q9KKvHU8eKi04wFAJ:-L4QXaiONgbLRo3IyU1H:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/resources.rs#L264)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Please use `AnimationControl` for the construction
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by torkleyy** [comment]
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[amethyst_animation/src/bundle.rs, line 93 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7dJWTmordw1cBMFg:-L4WIvIsOM8awVOUj_mc:bphycqa) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L93)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Released?
</blockquote></details>

Yep, https://docs.rs/shred/0.5.2/shred/struct.Resources.html#method.entry

---

*[amethyst_animation/src/bundle.rs, line 57 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4WJJBBjCziW27qvMFa:-L4WJJBBjCziW27qvMFb:bsojbvt) ([raw file](https://github.com/amethyst/amethyst/blob/6c1dc11f38aea75be2f9911bda6f8e1d4cbd3fb2/amethyst_animation/src/bundle.rs#L57)):*
> ```Rust
> ///
> /// Will add `SamplerInterpolationSystem<T>` with the given name.
> /// Will also add `SamplerProcessor<T::Scalar>` if it has not been added yet.
> ```

I think you misunderstood the usage of `add(sys, "", dependencies)`. It does not check if it's already added, it will just add the same system twice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by Rhuagh** [comment]
Review status: 6 of 19 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[amethyst_animation/src/bundle.rs, line 93 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4Q7dJWTmordw1cBMFg:-L4ZDBvsOiumfGH5C0fm:bb6un1j) ([raw file](https://github.com/amethyst/amethyst/blob/7375836ebe941c7e0dad600a0a6802b222ad07af/amethyst_animation/src/bundle.rs#L93)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Yep, https://docs.rs/shred/0.5.2/shred/struct.Resources.html#method.entry
</blockquote></details>

Done

---

*[amethyst_animation/src/bundle.rs, line 57 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4WJJBBjCziW27qvMFa:-L4WL3V1oVxFocBWhCpd:b-82t85u) ([raw file](https://github.com/amethyst/amethyst/blob/6c1dc11f38aea75be2f9911bda6f8e1d4cbd3fb2/amethyst_animation/src/bundle.rs#L57)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

I think you misunderstood the usage of `add(sys, "", dependencies)`. It does not check if it's already added, it will just add the same system twice.
</blockquote></details>

Ah, no, I just messed up the docs

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by Rhuagh** [comment]
Review status: 6 of 19 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[amethyst_animation/src/resources.rs, line 62 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4ZgVBjRz-YGojThL1q:-L4ZgVBjRz-YGojThL1r:bgp23mc) ([raw file](https://github.com/amethyst/amethyst/blob/adfa7aa3340d33ac8d83a32c0db410a40577ad26/amethyst_animation/src/resources.rs#L62)):*
> ```Rust
> /// Only required for animations which target more than a single node.
> #[derive(Debug, Clone)]
> pub struct AnimationHierarchy {
> ```

This should probably be generic on T aswell

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by torkleyy** [comment]
Reviewed 13 of 13 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[amethyst_animation/src/bundle.rs, line 57 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4WJJBBjCziW27qvMFa:-L4a6UKjRE9RwsxYAEn1:b2093ti) ([raw file](https://github.com/amethyst/amethyst/blob/6c1dc11f38aea75be2f9911bda6f8e1d4cbd3fb2/amethyst_animation/src/bundle.rs#L57)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Ah, no, I just messed up the docs
</blockquote></details>

Please correct it to "with the given name"

---

*[amethyst_animation/src/bundle.rs, line 97 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4a6d5e4EBiiqkS8stQ:-L4a6d5ffPBuejmmDOIN:b-c0d9g2) ([raw file](https://github.com/amethyst/amethyst/blob/adfa7aa3340d33ac8d83a32c0db410a40577ad26/amethyst_animation/src/bundle.rs#L97)):*
> ```Rust
>             .res
>             .entry()
>             .or_insert_with(|| AssetStorage::<Sampler<T::Primitive>>::new());
> ```

Can coerce the new function to a closure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

File: amethyst_animation/src/bundle.rs [hunk 1]
```diff
@@ -69,39 +82,54 @@ impl<'a> SamplingBundle<'a> {
...
+    ///
+    /// ### Parameters:
+    ///
+    /// - `animation_name`: name of the `AnimationControlSystem`
+    /// - `sampling_name`: name of the `SamplerInterpolationSystem`
+    pub fn new(animation_name: &'a str, sampling_name: &'a str) -> Self {
+        Self {
+            animation_name,
+            sampling_name,
+            dep: &[],
```
[the remark this answers is not in the document]
Comment (Rhuagh) [reply amethyst_animation/src/bundle.rs]: The `with_dep` function is hidden in the block below :)

[the remark this answers is not in the document]
Comment (Xaeroxe) [reply amethyst_animation/src/bundle.rs]: Gotcha, thanks, feel free to ignore.

**Comment by Rhuagh** [comment]
Review status: 15 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[amethyst_animation/src/bundle.rs, line 57 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4WJJBBjCziW27qvMFa:-L4m4oVuoehrVYKa8xsK:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/6c1dc11f38aea75be2f9911bda6f8e1d4cbd3fb2/amethyst_animation/src/bundle.rs#L57)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Please correct it to "with the given name"
</blockquote></details>

Done.

---

*[amethyst_animation/src/bundle.rs, line 97 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4a6d5e4EBiiqkS8stQ:-L4m4p06I_uuMso5eg3W:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/adfa7aa3340d33ac8d83a32c0db410a40577ad26/amethyst_animation/src/bundle.rs#L97)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Can coerce the new function to a closure.
</blockquote></details>

Done.

---

*[amethyst_animation/src/resources.rs, line 62 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4ZgVBjRz-YGojThL1q:-L4m4qCw6x3dLi3OWAvA:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/adfa7aa3340d33ac8d83a32c0db410a40577ad26/amethyst_animation/src/resources.rs#L62)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

This should probably be generic on T aswell
</blockquote></details>

Done.

---

*[amethyst_animation/src/resources.rs, line 72 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4I8sUNpWraGwOGoH6S-r3-72:-L4m4qkZzdXsFLwUAgg4:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/adfa7aa3340d33ac8d83a32c0db410a40577ad26/amethyst_animation/src/resources.rs#L72)):*
<details><summary><i>Previously, Xaeroxe (Jacob Kiesel) wrote…</i></summary><blockquote>

`targets` is the correct word here because this sentence is present tense.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by Rhuagh** [comment]
All comments addressed now I believe.

---

Review status: 15 of 19 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558#-:-L4m5AfUegXEK1alAiWb:b-anj4i8)*

File: amethyst_gltf/src/format/mod.rs [hunk 2]
```diff
@@ -208,44 +205,39 @@ fn load_channel(
...
-                },
+                LocalTransformChannel::Scale,
+                Sampler { input, ty, output },
             ))
         }
         Rotation => {
-            // gltf quat format: [x, y, z, w], our quat format: [w, x, y, z]
             let output = AccessorIter::<[f32; 4]>::new(sampler.output(), buffers)
-                .map(|q| [q[3], q[0], q[1], q[2]])
+                .map(|q| SamplerPrimitive::Vec4([q[3], q[0], q[1], q[2]]))
```
[the remark this answers is not in the document]
Comment (Rhuagh) [reply amethyst_gltf/src/format/mod.rs]: No, because it needs to be swizzled

[the remark this answers is not in the document]
Comment (Rhuagh) [reply amethyst_gltf/src/format/mod.rs]: Or hmm, it might

**Comment by Rhuagh** [comment]
Review status: 15 of 19 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[amethyst_gltf/src/systems.rs, line 234 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4NYdTObvxEKg2KVQXP-r4-234:-L4qs0qhktA33V7ATeAm:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/26dcb30d256d390d3f41307ad9988f6835f88c13/amethyst_gltf/src/systems.rs#L234)):*
<details><summary><i>Previously, omni-viral (Zakarum) wrote…</i></summary><blockquote>

Why `ref sampled` if it is cloned?
`sampler` now cloned two times.
</blockquote></details>

Done.

---

*[amethyst_gltf/src/format/mod.rs, line 187 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4NYdTObvxEKg2KVQXN-r4-187:-L4qs0Ss6O0ys4V6s3x5:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/26dcb30d256d390d3f41307ad9988f6835f88c13/amethyst_gltf/src/format/mod.rs#L187)):*
<details><summary><i>Previously, omni-viral (Zakarum) wrote…</i></summary><blockquote>

Consider to use `Result<Vec<_>, _>` and let the rest to be inferred.
</blockquote></details>

Done.

---

*[amethyst_gltf/src/format/mod.rs, line 229 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4NYdTObvxEKg2KVQXN-r4-229:-L4qs06ZY0-gWSs6kphO:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/26dcb30d256d390d3f41307ad9988f6835f88c13/amethyst_gltf/src/format/mod.rs#L229)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Or hmm, it might 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by Xaeroxe** [comment]
*[amethyst_animation/src/resources.rs, line 67 at r5](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4v9Dt6AxRgHW4wp9nq:-L4v9Dt6AxRgHW4wp9nr:bh0py4d) ([raw file](https://github.com/amethyst/amethyst/blob/5cec9de9c4da53b23c3eab76788e3c39099847bb/amethyst_animation/src/resources.rs#L67)):*
> ```Rust
> }
> 
> macro_rules! hashmap {
> ```

Cool macro!  We should consider moving this to `core`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558)*

**Comment by torkleyy** [comment]
:lgtm:

bors r+

---

Reviewed 3 of 4 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[amethyst_animation/src/resources.rs, line 67 at r5](https://reviewable.io:443/reviews/amethyst/amethyst/558#-L4v9Dt6AxRgHW4wp9nq:-L4vCBdtPGi0nbwEJJXk:b-4556gm) ([raw file](https://github.com/amethyst/amethyst/blob/5cec9de9c4da53b23c3eab76788e3c39099847bb/amethyst_animation/src/resources.rs#L67)):*
<details><summary><i>Previously, Xaeroxe (Jacob Kiesel) wrote…</i></summary><blockquote>

Cool macro!  We should consider moving this to `core`.
</blockquote></details>

There is some crate that exposes such a macro..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/558#-:-L4vCeNeRmfNI33dtWnI:b5abyu2)*

**Pull request merged**